test/simulation.jl

```diff
@@ -416,6 +416,81 @@ end
     end
 end
 
+@testset "SHAKE bond constraints" begin
+    n_atoms = 100
+    atom_mass = 10.0u"u"
+    atoms = [Atom(mass=atom_mass, σ=0.3u"nm", ϵ=0.2u"kJ * mol^-1") for i in 1:n_atoms]
+    boundary = CubicBoundary(2.0u"nm", 2.0u"nm", 2.0u"nm")
+
+    coords = place_atoms(n_atoms ÷ 2, boundary, min_dist=0.3u"nm")
+
+    for i in 1:length(coords)
+        push!(coords, coords[i].+ [0.15, 0.0, 0.0]u"nm")
+    end
+
+    temp = 100.0u"K"
+
+    velocities = [velocity(atom_mass, temp) for i in 1:n_atoms]
+
+    nb_matrix = trues(n_atoms, n_atoms)
+
+    for i in (1:n_atoms÷2)
+        nb_matrix[i, i + (n_atoms ÷ 2)] = false
+        nb_matrix[i + (n_atoms ÷ 2), i] = false
+    end
+
+    neighbor_finder = DistanceNeighborFinder(nb_matrix=nb_matrix, n_steps=10, dist_cutoff=1.5u"nm")
+    
+    bond_lengths = [0.1u"nm" for i in 1:(n_atoms÷2)]
+
+    sh = SHAKE(bond_lengths, collect(1:(n_atoms÷2)), collect((1+n_atoms÷2):n_atoms))
+    
+    constraint_list = (sh,)
+
+    sys = System(atoms=atoms,
+                 pairwise_inters=(LennardJones(nl_only=true),),
+                 constraints=constraint_list,
+                 coords=coords,
+                 velocities=velocities,
+                 boundary=boundary,
+                 neighbor_finder=neighbor_finder,
+                 loggers=Dict(
+                              "temp" => TemperatureLogger(10),
+                              "coords" => CoordinateLogger(10),
+                             )
+                )
+
+       
+    for i in 1:length(sys.coords)
+        sys.coords[i] += [rand()*0.01, rand()*0.01, rand()*0.01]u"nm"        
+    end
+
+    old_coords = sys.coords
+    apply_constraints!(sys, sh, old_coords, 0.002u"ps")
```

Comment:
We should probably test a case where iteration is required, i.e. where each atom can be in multiple constraints. Perhaps the test of a single `apply_constraints!` call could be done on a different set of atoms with a water-like bonding arrangement, i.e. constraints between 1-2, 2-3, 4-5, 5-6, etc. You could generate the coordinates with `place_atoms`, record the constraint distances from these coordinates, add random noise, call `apply_constraints!`, then check the constraint distances are matched.